File: Be-siukat-new/siukat-service-dev_snmptn_2022/services/ibu.js
var Ibu = require("../models/ibu");
var LogIbu = require("../models/log_ibu");
var Provinsi = require("../models/provinsi");
var Kabkot = require("../models/kabkot");
var Kecamatan = require("../models/kecamatan");
var Pekerjaan = require("../models/pekerjaan");

var ibu = function () { };

ibu.prototype.add = function (req, atribut) {
  return new Promise(function (resolve, reject) {
    Ibu.create(
      {
        no_peserta: req.no_peserta,
        status_ibu: req.status_ibu,
        nama_ibu: req.nama_ibu,
        nik_ibu: req.nik_ibu,
        telepon_ibu: req.telepon_ibu,
        alamat_ibu: req.alamat_ibu,
        provinsi_ibu: req.provinsi_ibu,
        kabkot_ibu: req.kabkot_ibu,
        kecamatan_ibu: req.kecamatan_ibu,
        pekerjaan_ibu: req.pekerjaan_ibu,
        penghasilan_ibu: req.penghasilan_ibu,
        sampingan_ibu: req.sampingan_ibu,
        scan_ktp_ibu: req.scan_ktp_ibu,
        scan_slip_ibu: req.scan_slip_ibu,
        tempat_lahir_ibu: req.tempat_lahir_ibu,
        tanggal_lahir_ibu: req.tanggal_lahir_ibu,
        atribut: atribut,
      },
      {
        fields: [
          "no_peserta",
          "status_ibu",
          "nama_ibu",
          "nik_ibu",
          "telepon_ibu",
          "alamat_ibu",
          "provinsi_ibu",
          "kabkot_ibu",
          "kecamatan_ibu",
          "pekerjaan_ibu",
          "penghasilan_ibu",
          "sampingan_ibu",
          "scan_ktp_ibu",
          "scan_slip_ibu",
          "tempat_lahir_ibu",
          "tanggal_lahir_ibu",
          "atribut",
        ],
      }
    )
      .then(function (response) {
        resolve(response);
      })
      .catch(function (err) {
        reject(err);
      });
  });
};

ibu.prototype.edit = function (req, no_peserta, atribut) {
  return new Promise(function (resolve, reject) {
    var data = {
      nama_ibu: req.body.nama_ibu,
      status_ibu: req.body.status_ibu,
      nik_ibu: req.body.status_ibu == "wafat" ? "" : req.body.nik_ibu,
      telepon_ibu: req.body.status_ibu == "wafat" ? "" : req.body.telepon_ibu,
      alamat_ibu: req.body.status_ibu == "wafat" ? "" : req.body.alamat_ibu,
      provinsi_ibu: req.body.status_ibu == "wafat" ? "" : req.body.provinsi_ibu,
      kabkot_ibu: req.body.status_ibu == "wafat" ? "" : req.body.kabkot_ibu,
      kecamatan_ibu:
        req.body.status_ibu == "wafat" ? "" : req.body.kecamatan_ibu,
      pekerjaan_ibu:
        req.body.status_ibu == "wafat" ? "" : req.body.pekerjaan_ibu,
      penghasilan_ibu:
        req.body.status_ibu == "wafat" ? "" : req.body.penghasilan_ibu,
      sampingan_ibu:
        req.body.status_ibu == "wafat" ? "" : req.body.sampingan_ibu,
      tempat_lahir_ibu:
        req.body.status_ibu == "wafat" ? "" : req.body.tempat_lahir_ibu,
      tanggal_lahir_ibu:
        req.body.status_ibu == "wafat" ? "" : req.body.tanggal_lahir_ibu,
      // scan_ktp_ibu: '',
      // scan_slip_ibu: '',
    };

    // Jika tidak bekerja
    // if (req.body.pekerjaan_ibu === "11") {
    //   data.penghasilan_ibu = "";
    //   data.sampingan_ibu = "";
    //   data.scan_slip_ibu = "";
    // }

    if (req.files["file_scan_ktp_ibu"] !== undefined) {
      data.scan_ktp_ibu =
        req.body.status_ibu == "wafat"
          ? ""
          : req.files["file_scan_ktp_ibu"][0].filename;
    }
    if (req.files["file_scan_slip_ibu"] !== undefined) {
      data.scan_slip_ibu =
        req.body.status_ibu == "wafat"
          ? ""
          : req.files["file_scan_slip_ibu"][0].filename;
    }

    Ibu.update(data, {
      where: {
        no_peserta: no_peserta,
        atribut: atribut,
      },
    })
      .then(function (response) {
        resolve(response);
      })
      .catch(function (err) {
        reject(err);
      });
  });
};

ibu.prototype.addLog = function (users, atribut, executor, timestamp) {
  return new Promise(function (resolve, reject) {
    LogIbu.create(
      {
        no_peserta: users.no_peserta,
        status_ibu: users.status_ibu,
        nama_ibu: users.nama_ibu,
        nik_ibu: users.nik_ibu,
        telepon_ibu: users.telepon_ibu,
        alamat_ibu: users.alamat_ibu,
        provinsi_ibu: users.provinsi_ibu,
        kabkot_ibu: users.kabkot_ibu,
        kecamatan_ibu: users.kecamatan_ibu,
        pekerjaan_ibu: users.pekerjaan_ibu,
        penghasilan_ibu: users.penghasilan_ibu,
        sampingan_ibu: users.sampingan_ibu,
        scan_ktp_ibu: users.scan_ktp_ibu,
        scan_slip_ibu: users.scan_slip_ibu,
        tempat_lahir_ibu: users.tempat_lahir_ibu,
        tanggal_lahir_ibu: users.tanggal_lahir_ibu,
        atribut: atribut,
        executor: executor,
        timestamp: timestamp,
      },
      {
        fields: [
          "no_peserta",
          "status_ibu",
          "nama_ibu",
          "nik_ibu",
          "telepon_ibu",
          "alamat_ibu",
          "provinsi_ibu",
          "kabkot_ibu",
          "kecamatan_ibu",
          "pekerjaan_ibu",
          "penghasilan_ibu",
          "sampingan_ibu",
          "scan_ktp_ibu",
          "scan_slip_ibu",
          "tempat_lahir_ibu",
          "tanggal_lahir_ibu",
          "atribut",
          "executor",
          "timestamp",
        ],
      }
    )
      .then(function (response) {
        resolve(response);
      })
      .catch(function (err) {
        reject(err);
      });
  });
};

ibu.prototype.getByLoggedIn = function (no_peserta) {
  return new Promise(function (resolve, reject) {
    Ibu.findOne({
      where: {
        no_peserta: no_peserta,
      },
      include: [Provinsi, Kabkot, Kecamatan, Pekerjaan],
    })
      .then(function (response) {
        resolve(response);
      })
      .catch(function (err) {
        reject(err);
      });
  });
};

ibu.prototype.checkData = function (no_peserta, users) {
  return new Promise((resolve, reject) => {
    Ibu.findOne({
      where: {
        no_peserta: no_peserta,
      },
      raw: true,
    })
      .then(function (response) {
        var result = true;
        var data = {};
        if (response.status_ibu === "wafat") {
          data.nama_ibu = response.nama_ibu;
        } else {
          if (users.ukt_tinggi === "ya") {
            delete response.nik_ibu;
            delete response.pekerjaan_ibu;
            delete response.penghasilan_ibu;
            delete response.scan_ktp_ibu;
            delete response.scan_slip_ibu;
          }

          // Jika tidak bekerja
          if (response.pekerjaan_ibu === "11") {
            // delete response.scan_slip_ibu
            delete response.penghasilan_ibu
          }

          delete response.sampingan_ibu;
          // delete response.telepon_ibu
          data = response;
        }

        for (var key in data) {
          result *= data[key] !== null && data[key] !== "" && data[key] !== 0;
        }
        resolve(result);
      })
      .catch(function (err) {
        reject(err);
      });
  });
};

module.exports = ibu;

File: BackEnd-Siukat/services/ibu.go
package services

import (
	"BackEnd-Siukat/config"
	"BackEnd-Siukat/models"
	"time"
)

type IbuService struct{}

// Add — parity dengan ibu.prototype.add di Node.js
func (s *IbuService) Add(req models.Ibu, atribut string) (models.Ibu, error) {
	db := config.DB
	req.Atribut = atribut
	err := db.Create(&req).Error
	return req, err
}

func (s *IbuService) Edit(data map[string]interface{}, noPeserta string, atribut string) (models.Ibu, error) {
	db := config.DB
	var ibu models.Ibu

	// SURGICAL FIX: Use Count instead of First to avoid "Scan error" on corrupted existing data
	var count int64
	db.Model(&models.Ibu{}).Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).Count(&count)

	if count == 0 {
		// 1. Jika BELUM ADA, maka Create
		data["no_peserta"] = noPeserta
		data["atribut"] = atribut
		if errCreate := db.Model(&models.Ibu{}).Create(data).Error; errCreate != nil {
			return models.Ibu{}, errCreate
		}
	} else {
		// 2. Jika SUDAH ADA, maka Update menggunakan Table/Where (aman dari Scan error)
		if errUpdate := db.Table("tb_ibu").Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).Updates(data).Error; errUpdate != nil {
			return models.Ibu{}, errUpdate
		}
	}

	// 3. Ambil data terbaru untuk return
	err := db.Preload("Pekerjaan").Preload("Provinsi").Preload("Kabkot").Preload("Kecamatan").
		Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&ibu).Error
	return ibu, err
}

// AddLog — Fix #4: Implementasi fungsi AddLog yang hilang
// Parity dengan ibu.prototype.addLog di Node.js
func (s *IbuService) AddLog(user models.Ibu, atribut string, executor string, timestamp *time.Time) error {
	db := config.DB
	logIbu := models.LogIbu{
		NoPeserta:       user.NoPeserta,
		StatusIbu:       user.StatusIbu,
		NamaIbu:         user.NamaIbu,
		NikIbu:          user.NikIbu,
		TeleponIbu:      user.TeleponIbu,
		AlamatIbu:       user.AlamatIbu,
		ProvinsiIbu:     user.ProvinsiIbu,
		KabkotIbu:       user.KabkotIbu,
		KecamatanIbu:    user.KecamatanIbu,
		PekerjaanIbu:    user.PekerjaanIbu,
		PenghasilanIbu:  user.PenghasilanIbu,
		SampinganIbu:    user.SampinganIbu,
		ScanKtpIbu:      user.ScanKtpIbu,
		ScanSlipIbu:     user.ScanSlipIbu,
		TempatLahirIbu:  user.TempatLahirIbu,
		TanggalLahirIbu: user.TanggalLahirIbu,
		Atribut:         atribut,
		Executor:        executor,
		Timestamp:       timestamp,
	}
	return db.Create(&logIbu).Error
}

// GetByLoggedIn — parity dengan ibu.prototype.getByLoggedIn di Node.js
func (s *IbuService) GetByLoggedIn(noPeserta string) (models.Ibu, error) {
	db := config.DB
	var ibu models.Ibu

	// ROBUST FETCH: Use the same strategy as AyahService
	// 1. PRIORITAS: Cek data 'sanggah' dulu dengan relasi lengkap
	err := db.Preload("Pekerjaan").Preload("Provinsi").Preload("Kabkot").Preload("Kecamatan").
		Where("no_peserta = ? AND atribut = ?", noPeserta, "sanggah").First(&ibu).Error
	
	if err == nil {
		return ibu, nil
	}

	// 2. FALLBACK: Jika tidak ada sanggah, ambil data 'original' dengan relasi lengkap
	err = db.Preload("Pekerjaan").Preload("Provinsi").Preload("Kabkot").Preload("Kecamatan").
		Where("no_peserta = ? AND atribut = ?", noPeserta, "original").First(&ibu).Error
	
	if err != nil {
		return models.Ibu{NoPeserta: noPeserta}, nil
	}

	return ibu, nil
}

// CheckData — parity dengan ibu.prototype.checkData di Node.js
func (s *IbuService) CheckData(noPeserta string, uktTinggi string) (bool, error) {
	db := config.DB
	var ibu models.Ibu
	if err := db.Where("no_peserta = ?", noPeserta).First(&ibu).Error; err != nil {
		return false, err
	}

	// Replicating Node.js "delete and check" logic using a map
	data := map[string]interface{}{
		"no_peserta":         ibu.NoPeserta,
		"status_ibu":       ibu.StatusIbu,
		"nama_ibu":         ibu.NamaIbu,
		"nik_ibu":          ibu.NikIbu,
		"telepon_ibu":      ibu.TeleponIbu,
		"alamat_ibu":       ibu.AlamatIbu,
		"provinsi_ibu":     ibu.ProvinsiIbu,
		"kabkot_ibu":       ibu.KabkotIbu,
		"kecamatan_ibu":    ibu.KecamatanIbu,
		"pekerjaan_ibu":    ibu.PekerjaanIbu,
		"penghasilan_ibu":  ibu.PenghasilanIbu,
		"sampingan_ibu":    ibu.SampinganIbu,
		"scan_ktp_ibu":     ibu.ScanKtpIbu,
		"scan_slip_ibu":    ibu.ScanSlipIbu,
		"tempat_lahir_ibu": ibu.TempatLahirIbu,
		"tanggal_lahir_ibu": ibu.TanggalLahirIbu,
		"atribut":           ibu.Atribut,
	}

	if ibu.StatusIbu == "wafat" {
		// Only name is required
		return ibu.NamaIbu != "", nil
	}

	if uktTinggi == "ya" {
		delete(data, "nik_ibu")
		delete(data, "pekerjaan_ibu")
		delete(data, "penghasilan_ibu")
		delete(data, "scan_ktp_ibu")
		delete(data, "scan_slip_ibu")
	}

	if ibu.PekerjaanIbu == "11" { // Tidak bekerja
		delete(data, "penghasilan_ibu")
	}

	delete(data, "sampingan_ibu")

	for _, val := range data {
		if val == nil || val == "" || val == 0 {
			return false, nil
		}
	}

	return true, nil
}
